### ISSUE TITLE
TooltipDataProcessor callback taint can break actionbars

### ISSUE THREAD

**Issue by Meorawr** [Stanzilla/WoWUIBugs#298]
The new TooltipDataProcessor system stores callbacks registered through the following FrameXML functions in a shared set of tables and, when dispatching them, does so in a way that taints execution. This can lead to situations where secure code that sets up a tooltip via an innocuous `GameTooltip:SetMeow()` call can become insecure and then break parts of the UI, including but not limited to actionbars.

```lua
function TooltipDataProcessor.AddTooltipPreCall(tooltipType, func) end
function TooltipDataProcessor.AddTooltipPostCall(tooltipType, func) end
function TooltipDataProcessor.AddLinePreCall(lineType, func) end
function TooltipDataProcessor.AddLinePostCall(lineType, func) end
```

As this system appears to be the replacement for the OnTooltipSetItem, OnTooltipSetSpell, and OnTooltipSetUnit script handlers having this _not_ break parts of the UI is somewhat desirable.

---

1. Run the following chat command: `/run TooltipDataProcessor.AddTooltipPostCall("ALL", nop)`.
2. Enter combat with an enemy.
3. Open the spellbook and hover over one of the displayed spell buttons.

At this point you'll receive an "Interface action failed because of an AddOn" message. Specifically, what's occurring here is the taint from [setting up the tooltip to show the spellbook entry leeches out to the action bar controller](https://github.com/Gethe/wow-ui-source/blob/15a65c6cfd23400d23c012c6ee4f1375e4b16abd/Interface/FrameXML/SpellBookFrame.lua#L610-L634), which will error on any of calls made to protected frames within. Additionally various fields and variables are modified, leading to this taint breaks cycling action bar pages while in-combat and using keybinds to activate actionbuttons.

A shortened version of the generated taint log from this specific interaction can be found below.

```
10/7 17:25:04.803  Execution tainted by *** ForceTaint_Strong *** while reading ALL - Interface/SharedXML/Tooltip/TooltipDataHandler.lua:34
10/7 17:25:04.803      Interface/SharedXML/Tooltip/TooltipDataHandler.lua:141
10/7 17:25:04.803      Interface/FrameXML/SpellBookFrame.lua:610
10/7 17:25:04.803  Global variable UpdateTooltip tainted by *** ForceTaint_Strong *** - Interface/FrameXML/SpellBookFrame.lua:611
10/7 17:25:04.803  Global variable ON_BAR_HIGHLIGHT_MARKS tainted by *** ForceTaint_Strong *** - Interface/FrameXML/ActionButton.lua:63 ClearOnBarHighlightMarks()
10/7 17:25:04.803      Interface/FrameXML/SpellBookFrame.lua:616
10/7 17:25:04.803  Global variable PET_ACTION_HIGHLIGHT_MARKS tainted by *** ForceTaint_Strong *** - Interface/FrameXML/PetActionBar.lua:30 PetActionBar:ClearPetActionHighlightMarks()
10/7 17:25:04.803      Interface/FrameXML/SpellBookFrame.lua:617
10/7 17:25:04.803  Global variable numForms tainted by *** ForceTaint_Strong *** - Interface/FrameXML/StanceBar.lua:23 StanceBar:Update()
10/7 17:25:04.803      Interface/FrameXML/ActionBarController.lua:128 ActionBarController_UpdateAll()
10/7 17:25:04.803      Interface/FrameXML/SpellBookFrame.lua:634
10/7 17:25:04.803  Execution tainted by *** ForceTaint_Strong *** while reading numForms - Interface/FrameXML/StanceBar.lua:24 StanceBar:Update()
10/7 17:25:04.803      Interface/FrameXML/ActionBarController.lua:128 ActionBarController_UpdateAll()
10/7 17:25:04.803      Interface/FrameXML/SpellBookFrame.lua:634
10/7 17:25:04.803  Interface/SharedXML/Tooltip/TooltipDataHandler.lua:34
10/7 17:25:04.803  An action was blocked in combat because of taint from *** ForceTaint_Strong *** - StanceButton1:SetShown()
10/7 17:25:04.803      Interface/FrameXML/ActionBar.lua:154 StanceBar:UpdateShownButtons()
10/7 17:25:04.803      Interface/FrameXML/StanceBar.lua:68 StanceBar:UpdateState()
10/7 17:25:04.803      Interface/FrameXML/StanceBar.lua:25 StanceBar:Update()
10/7 17:25:04.803      Interface/FrameXML/ActionBarController.lua:128 ActionBarController_UpdateAll()
10/7 17:25:04.803      Interface/FrameXML/SpellBookFrame.lua:634
10/7 17:25:04.803  Interface/SharedXML/Tooltip/TooltipDataHandler.lua:34
```
---

Possible solutions:

1. **Make use of a proper callback registry for the data and line post-call events.**
CallbackRegistryMixin already has all the provisions in place for dispatching isolated callbacks securely without tainting the entity that triggers an event. This system could also be used for pre-call events (with support for termination) if an intermediate closure were generated to handle setting a `terminated` flag in a temporary table during event dispatch.

```lua
-- The following is expositionary only.
function TooltipDataProcessor.AddTooltipPreCall(tooltipType, func) 
	local function PreCall(state, ...)
		if not state.terminated and func(...) then
			state.terminated = true
		end
	end;
	
	return TooltipCallbackRegistry:RegisterCallback(<something>, PreCall);
end
```

2. **Apply securecall to the `Process` function calls**.
This would be the simplest approach however - if done literally - wouldn't prevent taint from individual callbacks in the `calls` tables propagating between one another. This may cause issues if secure addons start registering for pre/post-call events through this mechanism after an insecure addon has registered any - as is already the case with the PTR feedback addon for example.

**Comment by Meorawr** [comment]
As an extension to the above, another test case and similar issue can be found below. This specific test case would only be applicable to beta and PTR clients as the taint in this case is caused by the PTR issue reporter.

1. Run the following command in the chat frame.
```
/run GameTooltip:SetOwner(UIParent, "ANCHOR_CURSOR") GameTooltip:SetText("Test") for i = 1, 10 do GameTooltip:AddLine("Test") end GameTooltip:Show()
```
2. Open the spellbook and hover over one of the displayed spell buttons.
3. Press any keybinding to use an ability on your actionbar.

You'll get an forbidden action popup which itself can be traced back to taint occurring when reading the `GameTooltipTextLeft<n>` regions within the tooltip post-call handler installed by the PTR issue reporter. An excerpt of the taint log can be found below. Note that either of the suggested resolutions in the initial part of this ticket would solve this issue implicitly, as a securecall boundary _somewhere_ in the processing of these handler functions would prevent the PTR issue reporter spreading taint back out of the system.

```
10/9 18:40:05.406  Execution tainted by *** ForceTaint_Strong *** while reading GameTooltipTextLeft8 - Interface/AddOns/Blizzard_PTRFeedback/Blizzard_PTRFeedback_Frames.lua:1096 HookIntoTooltip()
10/9 18:40:05.406      Interface/AddOns/Blizzard_PTRFeedback/Blizzard_PTRFeedback_Tooltips.lua:41 func()
10/9 18:40:05.406      Interface/SharedXML/Tooltip/TooltipDataHandler.lua:27 MakeCalls()
10/9 18:40:05.406      Interface/SharedXML/Tooltip/TooltipDataHandler.lua:37
10/9 18:40:05.406      Interface/SharedXML/Tooltip/TooltipDataHandler.lua:141
10/9 18:40:05.406      Interface/FrameXML/SpellBookFrame.lua:610

(Skipping PTR issue reporter taint; taint now spreads through SpellBookFrame to action bars)

10/9 18:40:05.406  Global variable UpdateTooltip tainted by *** ForceTaint_Strong *** - Interface/FrameXML/SpellBookFrame.lua:611
10/9 18:40:05.406  Global variable ON_BAR_HIGHLIGHT_MARKS tainted by *** ForceTaint_Strong *** - Interface/FrameXML/ActionButton.lua:63 ClearOnBarHighlightMarks()
10/9 18:40:05.406      Interface/FrameXML/SpellBookFrame.lua:616
10/9 18:40:05.406  Global variable PET_ACTION_HIGHLIGHT_MARKS tainted by *** ForceTaint_Strong *** - Interface/FrameXML/PetActionBar.lua:30 PetActionBar:ClearPetActionHighlightMarks()
10/9 18:40:05.406      Interface/FrameXML/SpellBookFrame.lua:617
10/9 18:40:05.406  Global variable numForms tainted by *** ForceTaint_Strong *** - Interface/FrameXML/StanceBar.lua:23 StanceBar:Update()
10/9 18:40:05.406      Interface/FrameXML/ActionBarController.lua:128 ActionBarController_UpdateAll()
10/9 18:40:05.406      Interface/FrameXML/SpellBookFrame.lua:634
10/9 18:40:05.406  Execution tainted by *** ForceTaint_Strong *** while reading numForms - Interface/FrameXML/StanceBar.lua:24 StanceBar:Update()
10/9 18:40:05.406      Interface/FrameXML/ActionBarController.lua:128 ActionBarController_UpdateAll()
10/9 18:40:05.406      Interface/FrameXML/SpellBookFrame.lua:634

(Skipping to use of actionbutton via keybind)

10/9 18:40:06.321  An action was blocked because of taint from *** ForceTaint_Strong *** - UseAction()
10/9 18:40:06.321      Interface/FrameXML/SecureTemplates.lua:364 handler()
10/9 18:40:06.321      Interface/FrameXML/SecureTemplates.lua:690 PerformAction()
10/9 18:40:06.321      Interface/FrameXML/SecureTemplates.lua:704 OnActionButtonClick()
10/9 18:40:06.321      Interface/FrameXML/SecureTemplates.lua:737 SecureActionButton_OnClick()
10/9 18:40:06.321      Interface/FrameXML/ActionButton.lua:104 TryUseActionButton()
10/9 18:40:06.321      Interface/FrameXML/ActionButton.lua:152 ActionButtonUp()
10/9 18:40:06.321      ACTIONBUTTON1:4
```

**Comment by Meorawr** [comment]
10.0.2.46144 made some changes that _somewhat_ help for insecurely-registered callbacks, but don't really solve the second issue the above comment. Additionally they introduce far more concerning issues that have been communicated back.

**Comment by Meorawr** [comment]
Fixed in some of the more recent 10.0.2 beta builds.
